pippy/graphsplit.py

```diff
@@ -0,0 +1,442 @@
+        )
+    )
+    assert allowed_node_imbalance >= 1.0 and allowed_edge_imbalance >= 1.0
+
+    # The optimization model contains N * K binary node variables such that
+    # x[i, j] = 1 if i-th node belongs to j-th stage, and x[i, j] = 0 otherwise.
+    num_node_vars = N * K
+    # Similarly, M * K binary edge variables x[i, j] = 1 if and only if both
+    # endpoints of the i-th edge belong to the j-th stage.
+    num_edge_vars = M * K
+    # An extra auxiliary variable for optimization.
+    num_aux_vars = 1
+    num_vars = num_node_vars + num_edge_vars + num_aux_vars
+
+    def node_var(node_idx: int, stage: int) -> int:
+        return node_idx * K + stage
+
+    def edge_var(edge_idx: int, stage: int) -> int:
+        return num_node_vars + edge_idx * K + stage
+
+    edge_aux_var = num_node_vars + num_edge_vars
+
+    # Define constraints for the optimization.
+    constraints = []
+
+    # Constraint 1:
+    #   - node/edge variables are synchronized with each other;
+    for i in range(M):
+        edge = edges[i]
+        for j in range(K):
+            # x[edge_var(i, j)] <= x[node_var(edge.source, j)]
+            A1 = zeros(num_vars)
+            A1[edge_var(i, j)] = 1
+            A1[node_var(edge.source, j)] = -1
+            constraints.append(LinearConstraint(A=A1, ub=0))
+            # x[edge_var(i, j)] <= x[node_var(edge.target, j)]
+            A2 = zeros(num_vars)
+            A2[edge_var(i, j)] = 1
+            A2[node_var(edge.target, j)] = -1
+            constraints.append(LinearConstraint(A=A2, ub=0))
+            # x[node_var(edge.source, j)] + x[node_var(edge.target, j)] - 1 <= x[edge_var(i, j)]
+            A3 = zeros(num_vars)
+            A3[node_var(edge.source, j)] = 1
+            A3[node_var(edge.target, j)] = 1
+            A3[edge_var(i, j)] = -1
+            constraints.append(LinearConstraint(A=A3, ub=1))
+
+    # Constraint 2:
+    #   - every node belongs to some stage;
+    for i in range(N):
+        A = zeros(num_vars)
+        for j in range(K):
+            A[node_var(i, j)] = 1
+        constraints.append(LinearConstraint(A=A, lb=1, ub=1))
+
+    # Constraint 3:
+    #   - every stage contains at least one edge;
+    for j in range(K):
+        A = zeros(num_vars)
+        for i in range(M):
+            A[edge_var(i, j)] = 1
+        constraints.append(LinearConstraint(A=A, lb=1))
+
+    # Constraint 4:
+    #   - edges go from a lower-index stage to an upper-index stage;
+    multiplier = [2 ** (K - j - 1) for j in range(K)]
+    for i in range(M):
+        edge = edges[i]
+        A = zeros(num_vars)
+        for j in range(K):
+            A[node_var(edge.target, j)] = multiplier[j]
+            A[node_var(edge.source, j)] = -multiplier[j]
+        constraints.append(LinearConstraint(A=A, ub=0))
+
+    # Constraint 5:
+    #   - nodes in every stage have (approximately) the same total weight;
+    sum_node_weights = sum(node.weight for node in nodes)
+    max_node_weight_per_stage = (
```

Comment:
Instead of doing it this way, I wanted to suggest an alternative way to control the imbalance (still within ILP framework). Why not directly constrain the max node/edge weight in any stage? Roughly we'd do it as follows:

Minimize alpha
subject to: node_weight_i_j <= alpha for i in all nodes in stage j

Of course this approach can be a bit more modular, and can be combined with other objectives. E.g. of we have some other (minimization) objective OBJ we can also just combine the two using:

Min OBJ + alpha
subject to: subject to node_weight_i_j <= alpha for i in all nodes in stage j

Etc.

I think the advantage of such an approach is that its a bit more "parameter free", i.e. we don't need an "allowed_node_imbalance" parameter.